runtime/memory.c

```diff
@@ -15,648 +15,347 @@
-   This is set when parsing [OCAMLRUNPARAM] and must stay constant
-   after that.
+/* Note [MM]: Enforcing the memory model.
+
+   Multicore OCaml implements the memory consistency model defined in
+
+     Bounding Data Races in Space and Time (PLDI '18)
+     Stephen Dolan, KC Sivaramakrishnan, Anil Madhavapeddy.
+
+   Unlike the C++ (also used in C11) memory model, this model gives
+   well-defined behaviour to data races, ensuring that they do not
+   affect unrelated computations. In C++, plain (non-atomic) accesses
+   have undefined semantics if they race, so it is necessary to use at
+   least relaxed atomics to implement all accesses.
+
+   However, simply using C++ relaxed atomics for non-atomic accesses
+   and C++ SC atomics for atomic ones is not enough, since the OCaml
+   memory model is stronger. The prototypical example where C++
+   exhibits a behaviour not allowed by OCaml is below. Assume that the
+   reference b and the atomic reference a are initially 0:
+
+       Thread 1            Thread 2
+       Atomic.set a 1;     let x = !b in
+       b := 1              let y = Atomic.get a in
+                           ...
+       Outcome: x = 1, y = 0
+
+   This outcome is not permitted by the OCaml memory model, as can be
+   seen from the operational model: if !b sees the write b := 1, then
+   the Atomic.set must have executed before the Atomic.get, and since
+   it is atomic the most recent set must be returned by the get,
+   yielding y = 1. In the equivalent axiomatic model, this would be a
+   violation of Causality.
+
+   If this example is naively translated to C++ (using atomic_{load,
+   store} for atomics, and atomic_{load, store}_explicit(...,
+   memory_order_relaxed) for nonatomics), then this outcome becomes
+   possible. The C++ model specifies that there is a total order on SC
+   accesses, but this total order is surprisingly weak. In this
+   example, we can have:
+
+       x = !b ...
+          [happens-before]
+       y = Atomic.get a
+          [SC-before]
+       Atomic.set a 1
+          [happens-before]
+       b := 1
+
+   Sadly, the composition of happens-before and SC-before does not add
+   up to anything useful, and the C++ model permits the read 'x = !b'
+   to read from the write 'b := 1' in this example, allowing the
+   outcome above.
+
+   To remedy this, we need to strengthen the relaxed accesses used for
+   non-atomic loads and stores. The most straightforward way to do
+   this is to use acquire loads and release stores instead of relaxed
+   for non-atomic accesses, which ensures that all reads-from edges
+   appear in the C++ synchronises-with relation, outlawing the outcome
+   above.
+
+   Using release stores for all writes also ensures publication safety
+   for newly-allocated objects, and isn't necessary for initialising
+   writes. The cost is free on x86, but requires a fence in
+   caml_modify on weakly-ordered architectures (ARM, Power).
```

Comment:
There is a difference between this description and what is described in the PLDI 2018 paper, isn't it? I.e. no release store for writes (i.e. `str` rather than `stlr`, no full `dmb` barrier).